In [85]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [86]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [87]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [88]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [89]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
scaled_data = StandardScaler().fit_transform(df_market_data.select_dtypes(include=['float64', 'int64']))

In [90]:
# Create a DataFrame with the scaled data
scaled_data_df = pd.DataFrame(scaled_data,
                              columns = df_market_data.select_dtypes(include=['float64', 'int64']).columns)

# Copy the crypto names from the original DataFrame
scaled_data_df["coin_id"] = df_market_data.index

# Set the coin_id column as index
scaled_data_df.set_index("coin_id", inplace=True)

# Display the scaled DataFrame
scaled_data_df.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [91]:
# Create a list with the number of k-values from 1 to 11

k = list(range(1, 11))

In [92]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
for i in k:
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
    k_model = KMeans(n_clusters=i, random_state=1)
# 2. Fit the model to the data using `df_market_data_scaled`
    k_model.fit(scaled_data_df)
# 3. Append the model.inertia_ to the inertia list
    inertia.append(k_model.inertia_)
    print(f"for k={i}, we have an inertia of {k_model.inertia_}")

for k=1, we have an inertia of 287.0
for k=2, we have an inertia of 212.1233420748626
for k=3, we have an inertia of 165.13675233446816
for k=4, we have an inertia of 79.02243535120977
for k=5, we have an inertia of 66.41305075939437
for k=6, we have an inertia of 61.6340881106822
for k=7, we have an inertia of 46.074467441030215
for k=8, we have an inertia of 43.39269411734827
for k=9, we have an inertia of 32.91322265971525
for k=10, we have an inertia of 30.46911129327036


c:\Users\bryan\anaconda3\envs\dev\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\bryan\anaconda3\envs\dev\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\bryan\anaconda3\envs\dev\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\bryan\anaconda3\envs\dev\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is k

In [93]:
# Create a dictionary with the data to plot the Elbow curve
elbow_df = pd.DataFrame({"k": k, "inertia": inertia})

# Create a DataFrame with the data to plot the Elbow curve
elbow_df.head()

k     inertia
0  1  287.000000
1  2  212.123342
2  3  165.136752
3  4   79.022435
4  5   66.413051

In [94]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.

elbow_plot_original = elbow_df.hvplot.line(x="k", y="inertia")
elbow_plot_original

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** The best value for 'k' is 4.


---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [95]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=1)

In [96]:
# Fit the K-Means model using the scaled DataFrame
model.fit(scaled_data_df)

c:\Users\bryan\anaconda3\envs\dev\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=1)

In [97]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
k_4 = model.predict(scaled_data_df)

# Print the resulting array of cluster values.
print(k_4)

[2 2 0 0 2 2 2 2 2 0 0 0 0 2 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 0 0 0 3 2 0 0 1
 0 0 0 0]


In [98]:
# Create a copy of the scaled DataFrame
scaled_data_df = scaled_data_df.copy()

In [99]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
scaled_data_df['cluster'] = k_4

# Display the copy of the scaled DataFrame
scaled_data_df.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  cluster  
coin_id                                            
bitcoin                        -0.251637        2  
ethereum                       -0.199352        2  
tether                         -0.282061        0  
ripple                         -0.295546        0  
bitcoin-cash                   -0.270317        2

In [100]:
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.

# Reset the index to make 'coin_id' a column
#scaled_data_df = scaled_data_df.reset_index()

cluster_plot_original = scaled_data_df.hvplot.scatter(
    x= "price_change_percentage_24h",
    y= "price_change_percentage_7d",
    by= "cluster",
    hover_cols= ["coin_id"]
)
cluster_plot_original

:NdOverlay   [cluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [101]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [102]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components.
scaled_pca = pca.fit_transform(scaled_data_df)

# View the scaled PCA data
scaled_pca[:5]

array([[ 0.44890795,  1.24537573,  0.8506404 ],
       [ 0.49536709,  0.899823  ,  1.31755943],
       [-0.81884571, -0.07189909, -0.69501507],
       [-0.84035711, -0.08005407, -0.54435952],
       [ 0.81324004,  2.66952035,  1.64332113]])

In [103]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
pca.explained_variance_ratio_


array([0.34871677, 0.31363391, 0.22627118])

In [104]:
# Calculate the total explained variance of the three principal components
total_explained_variance = sum(pca.explained_variance_ratio_)
total_explained_variance

0.8886218549859446

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** The total explained variance is 0.8886218549859446.

In [105]:
# Create a new DataFrame with the PCA data.
scaled_pca_df = pd.DataFrame(
    scaled_pca,
    columns=["PC1", "PC2", "PC3"]
)

# Copy the crypto names from the original scaled DataFrame
scaled_pca_df["coin_id"] = scaled_data_df.index

# Set the coin_id column as index
scaled_pca_df.set_index("coin_id", inplace=True)

# Display the scaled PCA DataFrame
scaled_pca_df

PC1       PC2       PC3
coin_id                                           
bitcoin               0.448908  1.245376  0.850640
ethereum              0.495367  0.899823  1.317559
tether               -0.818846 -0.071899 -0.695015
ripple               -0.840357 -0.080054 -0.544360
bitcoin-cash          0.813240  2.669520  1.643321
binancecoin           0.822871  1.682096  0.905203
chainlink             0.913868  1.128684  2.730038
cardano               0.811102  1.113378  1.598655
litecoin              0.258728  0.946884  1.103560
bitcoin-cash-sv      -1.003921  0.119410 -0.278564
crypto-com-chain     -1.363289 -1.245349 -0.950712
usd-coin             -0.830643 -0.077047 -0.707661
eos                  -1.135458 -0.171401 -0.441569
monero                2.084978  2.688979  0.776469
tron                 -0.633198  0.016911 -0.249426
tezos                -0.161908  0.310904  1.818521
okb                  -0.996987 -1.292574 -0.793701
stellar              -0.990512 -0.449877 -0.052002
cosmos                0.878129  1.093136  1.774115
cdai                 -0.874815 -0.008933 -0.694993
neo                  -1.135199 -0.792113 -0.499484
wrapped-bitcoin       0.434801  1.232506  0.839794
leo-token            -1.110348 -0.540593 -0.854934
huobi-token          -0.758629 -0.094085 -0.466658
nem                   0.222636 -0.083977 -1.584956
binance-usd          -0.826142 -0.049070 -0.700753
iota                 -1.059844 -0.097044  0.149781
vechain              -1.808829 -1.520236 -0.527335
zcash                 0.558082  0.920479  1.377145
theta-token           1.452903 -1.756424 -2.506866
dash                 -0.897830 -0.134758  0.197015
ethereum-classic     -0.908899 -0.095612 -0.194364
ethlend               6.049082 -7.033268  2.235723
maker                 0.307206  0.565760  0.968488
havven               -0.507111 -2.407910  0.392697
omisego              -0.236488  0.071399 -1.445063
celsius-degree-token  6.454065  2.823519 -4.810646
ontology             -1.823588 -1.561754 -0.000663
ftx-token            -0.689158  0.253900 -0.073233
true-usd             -0.823298 -0.031016 -0.694746
digibyte             -0.770668 -0.187671 -0.911020

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [106]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1, 11))

In [107]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
for i in k:
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
    k_model = KMeans(n_clusters=i, random_state=1)
# 2. Fit the model to the data using `df_market_data_pca`
    k_model.fit(scaled_pca_df)
# 3. Append the model.inertia_ to the inertia list
    inertia.append(k_model.inertia_)
    print(f"for k={i}, we have an inertia of {k_model.inertia_}")


for k=1, we have an inertia of 290.6226939879642
for k=2, we have an inertia of 207.30743902517526
for k=3, we have an inertia of 123.12695056327829
for k=4, we have an inertia of 44.130602253217134
for k=5, we have an inertia of 33.2174272551524
for k=6, we have an inertia of 23.374713031741738
for k=7, we have an inertia of 20.943077394523403
for k=8, we have an inertia of 18.666961001043948
for k=9, we have an inertia of 13.640110276427894
for k=10, we have an inertia of 10.287149461797636


c:\Users\bryan\anaconda3\envs\dev\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\bryan\anaconda3\envs\dev\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\bryan\anaconda3\envs\dev\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\bryan\anaconda3\envs\dev\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is k

In [108]:
# Create a dictionary with the data to plot the Elbow curve
pca_elbow_data = {"k": k, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve
pca_df_elbow = pd.DataFrame(pca_elbow_data)

In [109]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.

elbow_plot_pca = pca_df_elbow.hvplot.line(
    x="k",
    y="inertia",
    title="Elbow Curve",
    xticks=k
)
elbow_plot_pca

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** The best value for 'k' is 4 when using the PCA data.


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** This does not differ from the best k value found using the original data.

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [110]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=0)

In [111]:
# Fit the K-Means model using the PCA data
model.fit(scaled_pca_df)

c:\Users\bryan\anaconda3\envs\dev\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=0)

In [112]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
pca_k_4 = model.predict(scaled_pca_df)

# Print the resulting array of cluster values.
print(pca_k_4)

[2 2 0 0 2 2 2 2 2 0 0 0 0 2 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 0 0 0 1 2 0 0 3
 0 0 0 0]


In [113]:
# Create a copy of the scaled PCA DataFrame
scaled_pca_predictions_df = scaled_pca_df.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
scaled_pca_predictions_df["predicted clusters"] = pca_k_4

# Display the copy of the scaled PCA DataFrame
scaled_pca_predictions_df.head()

PC1       PC2       PC3  predicted clusters
coin_id                                                       
bitcoin       0.448908  1.245376  0.850640                   2
ethereum      0.495367  0.899823  1.317559                   2
tether       -0.818846 -0.071899 -0.695015                   0
ripple       -0.840357 -0.080054 -0.544360                   0
bitcoin-cash  0.813240  2.669520  1.643321                   2

In [114]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.

cluster_plot_pca = scaled_pca_predictions_df.hvplot.scatter(
    x= "PC1",
    y= "PC2",
    by= "predicted clusters",
    hover_cols= ["coin_id"]
)
cluster_plot_pca

:NdOverlay   [predicted clusters]
   :Scatter   [PC1]   (PC2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [115]:
# Composite plot to contrast the Elbow curves
composite_plot_elbow = elbow_plot_original + elbow_plot_pca
composite_plot_elbow


:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [116]:
# Composite plot to contrast the clusters
composite_plot_clusters = cluster_plot_original + cluster_plot_pca
composite_plot_clusters

:Layout
   .NdOverlay.I  :NdOverlay   [cluster]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [predicted clusters]
      :Scatter   [PC1]   (PC2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** The impact of using fewer features to cluster the data using K-Means is that the inertia changes. In the left figure, there is more distance between each data point in each cluster, whereas in the right figure there is less distance between each data point in each cluster. This shows us that when we cluster the data using K-Means, the inertia is higher than when we use more features to cluster the data.